maven-embedder/src/test/java/org/apache/maven/cli/MavenCliTest.java

```diff
@@ -364,6 +369,38 @@ public void testVersionStringWithoutAnsi() throws Exception {
         assertEquals(MessageUtils.stripAnsiCodes(versionOut), versionOut);
     }
 
+    @Test
+    public void testPropertiesInterpolation() throws Exception {
+        // Arrange
+        CliRequest request = new CliRequest(
+                new String[] {
+                    "-Dfoo=bar",
+                    "-DvalFound=s${foo}i",
+                    "-DvalNotFound=s${foz}i",
+                    "-DvalRootDirectory=${session.rootDirectory}/.mvn/foo",
+                    "-DvalTopDirectory=${session.topDirectory}/pom.xml",
```

Comment:
I have a problem with these two expressions: They haven't been back ported to 3.9.x, no? So they are fake here and not work throughout the entire build?

Reply:
Expressions HAVE BEEN (in this PR) backported, what you *really mean* is that 3.9.x does not expose them on MavenSession object, that is true. Still, the point of this PR is to make builds between 3.9.x and 4 permeable, "work with both". This PR as explained above, *does not back port everything*, just the ability to interpolate and these two "special" expressions used in interpolation.

Reply:
This I understand now. Hope they won't be misunderstood by users.